mesonbuild/backend/backends.py

```diff
@@ -197,15 +197,20 @@ def flatten_object_list(self, target, proj_dir_to_build_root=''):
 
     def serialise_executable(self, exe, cmd_args, workdir, env={},
                              capture=None):
-        import uuid
+        import hashlib
         # Can't just use exe.name here; it will likely be run more than once
         if isinstance(exe, (dependencies.ExternalProgram,
                             build.BuildTarget, build.CustomTarget)):
             basename = exe.name
         else:
             basename = os.path.basename(exe)
-        scratch_file = 'meson_exe_{0}_{1}.dat'.format(basename,
-                                                      str(uuid.uuid4())[:8])
+        # Take a digest of the cmd args, env, workdir, and capture. This avoids
+        # collisions and also makes the name deterministic over regenerations
+        # which avoids a rebuild by Ninja because the cmdline stays the same.
+        data = bytes(str(env) + str(cmd_args) + str(workdir) + str(capture),
```

Comment:
The order in which key/value pairs are popped from a dict is unspecified but is stable. For all known versions of Python, if you insert items in two dictionaries in a specific order, the items will pop in a (different) specific order. You can check this yourself, and the unit test also checks for this.

In fact, starting from 3.6, it is specified to always [return in the same order](https://mail.python.org/pipermail/python-dev/2016-September/146327.html) in which the items were inserted, so we will always be able to rely on this.